src/index.ts

```diff
@@ -0,0 +1,149 @@
+import {
+  requestsCount,
+  requestDuration,
+  requestStageDuration,
+} from "./metrics";
+import { parseUrlTemplate } from "./utils";
+
+interface Headers {
+  [header: string]: number | string | string[] | undefined;
+}
+
+interface Options {
+  baseUrl: string;
+  headers: Headers;
+}
+
+interface Timings {
+  start?: number;
+  socket?: number;
+  lookup?: number;
+  connect?: number;
+  upload?: number;
+  response?: number;
+  end?: number;
+  error?: number;
+  phases: {
+    wait?: number;
+    dns?: number;
+    tcp?: number;
+    request?: number;
+    firstByte?: number;
+    download?: number;
+    total?: number;
+  };
+}
+
+interface Result {
+  timings?: Timings;
+  statusCode?: number;
+  body?: unknown;
+}
+
+type Method =
+  | "GET"
+  | "PUT"
+  | "POST"
+  | "PATCH"
+  | "DELETE"
+  | "HEAD"
+  | "OPTIONS"
+  | "TRACE"
+  | "CONNECT";
+
+export class HTTPClient {
```

Comment:
So for tests you would want to:

1) create a simple HTTP server that has a few endpoints, eg `GET/POST /things` and `GET/DELETE /things/{id}` that returns/errors with certain values

2) create a TestHttpClient that extends this class and uses these endpoints

3) call it with different values, check for expected success and error conditions